CMakeLists.txt

```diff
@@ -434,6 +434,17 @@ message(STATUS "Runtime checking level: CHECK=${CHECK}")
 target_compile_definitions(bout++ PUBLIC "CHECK=${CHECK}")
 set(BOUT_CHECK_LEVEL ${CHECK})
 
+
+if (CHECK GREATER 1)
+  set(default ON)
+else()
+  set(default OFF)
+endif()
+set(BOUT_USE_MSGSTACK ${default} CACHE STRING "Set the Mesagestack level" )
+set_property(CACHE BOUT_USE_MSGSTACK PROPERTY TYPE BOOL)
+message(STATUS "Message stack: BOUT_USE_MSGSTACK=${BOUT_USE_MSGSTACK}")
+target_compile_definitions(bout++ PUBLIC "BOUT_USE_MSGSTACK=${BOUT_USE_MSGSTACK}")
```

Comment:
Now that I come to look at this properly, I see the [`cmake_dependent_option`](https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html) immediately below should do exactly what we want :facepalm: 

```suggestion
cmake_dependent_option(BOUT_ENABLE_MSGSTACK "Enable debug message stack" ON
  "CHECK GREATER 1" OFF)
message(STATUS "Message stack: BOUT_USE_MSGSTACK=${BOUT_ENABLE_MSGSTACK}")
set(BOUT_USE_MSGSTACK ${BOUT_ENABLE_MSGSTACK})
```

What this does is: if `CHECK > 1`, then present the option to turn `MsgStack` on/off, but defaults to on. Otherwise, turn it off.

What this doesn't do is allow the user to turn `MsgStack` on for `CHECK <= 1` -- it's always off in that case.

If we do actually want complete control, then the current method like this works fine, although I would make the following changes:

```suggestion
if (CHECK GREATER 1)
  set(bout_use_msgstack_default ON)
else()
  set(bout_use_msgstack_default OFF)
endif()
set(BOUT_ENABLE_MSGSTACK ${bout_use_msgstack_default} CACHE BOOL "Enable debug message stack")
message(STATUS "Message stack: BOUT_USE_MSGSTACK=${BOUT_ENABLE_MSGSTACK}")
set(BOUT_USE_MSGSTACK ${BOUT_ENABLE_MSGSTACK})
```

The naming convention is `BOUT_ENABLE_*` for the cmake option, and then `BOUT_USE_*` for the `#define`. Third-party dependencies are `BOUT_USE` for the option, `BOUT_HAS` for the define. I can't really remember why I settled on this particular naming convention, historical reasons I suspect. 

`target_compile_definitions` isn't needed, it's all handled through the `build_defines.hxx` header.

Reply:
I think it might be useful to keep that options - with optimisations the formatting should be much faster - and for large meshes probably negligible.